File: c++/cote/GJM/강결합_컴포넌트_분리.cpp
#include <iostream>
#include <vector>
#include <random>
#include <functional>
#include <algorithm>
#include <stack>
#define TEST_CASES 10
#define N 1000

using namespace std;


void _dfs(int cur, vector<bool>& visited, vector<vector<int>>& graph) {
	visited[cur] = true;

	for (auto& next : graph[cur]) {
		if (!visited[next]){
			_dfs(next, visited, graph);
		}
	}
}

bool check_reachable(vector<vector<int>>& graph) {
	vector<bool> visited(graph.size(), false);

	_dfs(0, visited, graph);

	return all_of(begin(visited), end(visited), [](bool v){return v;});
}

vector<vector<int>> initialize(int n, function<bool()> edge_gen) {
	vector<vector<int>> ret;

	while(true) {

		ret = vector(n, vector<int>());

		for (int i = 0; i < n; i++) {
			for (int j = 0; j < n; j++) {
				if (i == j || !edge_gen()) {
					continue;
				}

				ret[i].push_back(j);
			}
		}

		if (check_reachable(ret)) {
			break;
		}
	}

	return ret;
}

int dfs(int cur, vector<vector<int>>& graph, vector<bool>& finished, vector<int>& discovered, vector<int>& sccIds, stack<int>& st, int& vertexCounter, int& sccCounter) {
	
	discovered[cur] = vertexCounter++;
	int ret = discovered[cur];
	
	st.push(cur);

	for (const auto& next : graph[cur]) {
		if (discovered[next] == -1) {
			int subtree = dfs(next, graph, finished, discovered, sccIds, st, vertexCounter, sccCounter);
			ret = min(ret, subtree);
		} else if (discovered[next] < discovered[cur] && !finished[next] ) {
			ret = min(ret, discovered[next]);
		} 
	}

	if (ret == discovered[cur]) {
		while (true) {
			int t = st.top();
			st.pop();
			sccIds[t] = sccCounter;
			if  (t == cur) break;
		}	
		sccCounter++;
	}

	finished[cur] = true;

	return ret;
}

vector<int> solve(vector<vector<int>>& graph) {
	vector<int> sccIds = vector<int>(graph.size(), -1);
	vector<bool> finished(graph.size(), false);
	vector<int> discovered(graph.size(), -1);
	stack<int> st;

	int vertexCounter = 0;
	int sccCounter = 0; 

	dfs(0, graph, finished, discovered, sccIds, st, vertexCounter, sccCounter);

	return sccIds;
}

int main(void) {
	default_random_engine gen(43);
	auto n_gen {bind(uniform_int_distribution{3, N}, gen)};
	auto edge_gen {bind(bernoulli_distribution{0.05}, gen)};

	for (int it = 0; it < TEST_CASES; it++) {
		int n = n_gen(gen);
		vector<vector<int>> adj = initialize(n, edge_gen);		

		vector<int> sccIds = solve(adj);
		for (const auto& sccId : sccIds) {
			cout << sccId << " ";
		}
		cout << endl << endl;
	}

	return 0;
}

File: go/cote/GJM/강결합_컴포넌트_분리.go
package main

import (
	"container/list"
	"fmt"
	"math/rand"
	"slices"
)

const (
	TEST_CASES = 10
	N          = 1000
)

func _dfs(cur int, adj [][]int, visited []bool) {
	visited[cur] = true

	for _, next := range adj[cur] {
		if !visited[next] {
			_dfs(next, adj, visited)
		}
	}
}

func check_reachable(adj [][]int) bool {
	visited := make([]bool, len(adj))

	_dfs(0, adj, visited)

	return !slices.Contains(visited, false)
}

func initialize(random *rand.Rand) (ret [][]int) {
	n := random.Intn(N-3) + 3
	for {
		ret = make([][]int, n)

		for i, _ := range ret {
			for j := 0; j < n; j++ {
				if i == j || random.Float32() < 0.975 {
					continue
				}
				ret[i] = append(ret[i], j)
			}
		}

		if check_reachable(ret) {
			break
		}
	}

	return ret
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

func dfs(cur int, finished []bool, discovered, sccIds []int, stack *list.List, vertex_counter, scc_counter *int, adj [][]int) int {
	discovered[cur] = *vertex_counter
	(*vertex_counter) += 1
	ret := discovered[cur]

	stack.PushBack(cur)

	for _, next := range adj[cur] {
		if discovered[next] == -1 {
			subtree := dfs(next, finished, discovered, sccIds, stack, vertex_counter, scc_counter, adj)
			ret = min(ret, subtree)
		} else if discovered[next] < discovered[cur] && !finished[next] {
			ret = min(ret, discovered[next])
		}
	}

	if ret == discovered[cur] {
		for {
			t := stack.Back().Value.(int)
			stack.Remove(stack.Back())
			sccIds[t] = *scc_counter

			if t == cur {
				break
			}
		}
		(*scc_counter) += 1
	}

	finished[cur] = true

	return ret
}

func solve(adj [][]int) []int {
	n := len(adj)
	finished := make([]bool, n)
	discovered := make([]int, n)
	sccIds := make([]int, n)
	stack := list.New()
	vertex_counter, scc_counter := 0, 0

	for i, _ := range finished {
		finished[i] = false
	}

	for i, _ := range discovered {
		discovered[i] = -1
	}

	for i, _ := range sccIds {
		sccIds[i] = -1
	}

	dfs(0, finished, discovered, sccIds, stack, &vertex_counter, &scc_counter, adj)

	return sccIds
}

func main() {
	random := rand.New(rand.NewSource(43))

	for it := 0; it < TEST_CASES; it++ {
		adj := initialize(random)
		fmt.Println(solve(adj))
	}
}
